README.rst
| Code_ | Bugs_ | Forum_ | Docs_ | License_ | Contact_

.. _Code : http://github.com/frgomes/sphinx_typesafe
.. _Bugs : http://github.com/frgomes/sphinx_typesafe/issues
.. _Forum : http://github.com/frgomes/sphinx_typesafe/wiki
.. _Docs : http://sphinx_typesafe.readthedocs.org
.. _License : http://opensource.org/licenses/Apache-2.0
.. _Contact : http://github.com/~frgomes



``sphinx_typesafe`` is a decorator which enables dynamic type checking on Python 
method and function calls. It works in conjunction with `Sphinx-style docstrings`_,
which makes it particularly convenient for keeping the code documentation up-to-date
with the code actually being executed.

.. _`Sphinx-style docstrings`: http://sphinx-doc.org/markup/desc.html#info-field-lists


Features is a Nutshell
======================

* The decorator can be attached to any function or method.

* Raises ``TypeError`` if types of arguments do not match the specification.

* Raises ``TypeError`` if type of return value does not match the specification.

* Performs dynamic type checking.


Python2
=======

Since function annotations are not available in Python2 the way type checking for Python2 is a documentation convention for parameters based on `the info field lists of sphinx`_. So even when you don't use type checking you can use it to generate documentation.

.. _`the info field lists of sphinx`: http://sphinx-doc.org/markup/desc.html#info-field-lists


Syntax for Python2 using sphinx style docstrings
------------------------------------------------

This is the preferred way since you will be also documenting your code.

::

	@typesafe
	def foo(param_a, param_b):
		"""
		:type param_a: 	types.StringType
		:type param_b: 	types.IntType
		:rtype:         types.BooleanType	
		"""
		# Do Something 
		return True


.. note::

    Observe the usage of ``rtype`` to specify the type returned by the function.



Syntax for Python2 using decorator arguments
--------------------------------------------

This is an alternative approach, useful in circunstances where Sphinx-style documentation is not allowed or desired, for whatever reason.

::

	@typesafe( { "param_a" : str, 
		     "param_b" : types.IntType, 
		     "param_c" : own_module.OwnType
		     "return"  : bool }
		     )
	def foo(param_a, param_b, param_c):
		""" Some Docstring Info		 """
		# Do Something 
		return True

.. note::

   Observe the usage of ``return`` to specify the type returned by the function.



You can use any Python type
---------------------------

So if you have defined a ``Point`` class in module ``mod1`` like below:

::

    # File: mod1.py

    class Point(object):
	def __init__(self, x = None, y = None):
            """ Initialize the Point. Can be used to give x,y directly."""
	    self.x = x
	    self.y = y

then you can employ this type in your code like this:

::

   from mod1 import Point

   @typesafe
   def foo(afunc):
       """ 
       :type afunc: 	mod1.Point
       :rtype: 		types.BooleanType
       """
       return True


Python3
=======

.. warning::

    This is a tentative implementation which is not finished at the moment!!


The base technique is the Function Annotations proposed in `PEP-3107`_ which is 
documented in `Python3 What's New`_ (see section New Syntax).


.. _`PEP-3107`: http://www.python.org/dev/peps/pep-3107
.. _`Python3 What's New`: http://docs.python.org/3.0/whatsnew/3.0.html


Syntax for Python3
------------------

::

	@typesafe
	def foo(param_a: str, param_b: int) -> bool:
		# Do Something 
		return True


* The @typesafe decorator will then check all arguments dynamically whenever the foo is called for valid types.

* As a quoting remark from the PEP 3107: "All annotated parameter types can be any python expression.", but for typechecking only types make sense, though.

The idea and parts of the implementation were inspired by the book: `Pro Python (Expert's Voice in Open Source)`_

.. _`Pro Python (Expert's Voice in Open Source)`: http://www.amazon.com/Python-Experts-Voice-Open-Source/dp/1430227575



FAQ
===

Why it was called IcanHasTypeCheck ?
------------------------------------

*IcanHasTypeCheck (ICHTC)*, refers to the `famous lolcats`_.

.. _`famous lolcats`: http://en.wikipedia.org/wiki/I_Can_Has_Cheezburger%3F


Why is now called sphinx_typesafe ?
-----------------------------------

Because *typesafe* tells immediatelly what it is about. Unfortunately, *typesafe* was already taken on PyPI, so *sphinx_typesafe* seemed to be a good alternative name which also relates to the documentation standard adopted.


Support
=======

Please find links on the top of this page.


CHANGES.rst
CHANGES
=======

0.2 (03-dec-2013)
-----------------

* Full code rewiew and rewrite by Richard Gomes <rgomes.info@gmail.com>

    * @typesafe is now a class

    * Code reorganization and some optimization

    * Preparing code for porting to Python3

* New features

    * Added verification of return type of decorated functions

* Bugfixes

    * Type resolver now finds types involving multiple nested modules

    * Type resolver assumes module '__builtin__' for simple names, such as 'bool'

* Renamed to *sphinx_typesafe*

* Documentation rewriten in reStructuredText

* Added *docs* directory required by readthedocs.org

* First version deployed onto PyPI


0.1
---

Original version of IcanHasTypeCheck (ICHTC) by Klaas <khz@tzi.org>


tests/mod1.py
#
# example Module for testing @typesafe
# 

class Point(object):
	
	def __init__(self, x = None, y = None):
		""" Initialize the Point. Can be used to give x,y directly."""
		self.x = x
		self.y = y

	def set_coord(self, x , y):
		self.x = x
		self.y = y		

		


tests/i_is_example.py
#
# Example Script for IcanHasType  
# typecheck library for python 
# 


# import re, sys
# import inspect
from sphinx_typesafe.typesafe import typesafe
import mod1


@typesafe
def foo(param_a, param_b):
	""" 
		:type param_a: 	types.IntType
		:type param_b: 	types.StringType
		:rtype:         types.BooleanType	
	"""
	# Do Something 
	print "foo"
	return True


#{ "param_a" : "mod1.Point"}
@typesafe({"param_a": str})
def mod_test( param_a ):
	"""
		:type param_a: mod1.Point
	"""	
	print "Yeah, I was called"


if __name__ == "__main__":
	# passes successfully (see function  spec above)
	foo(1, "test")

	# fails 'successfully' (see function  spec above)
	#foo(1, 2)
	p = mod1.Point()
	mod_test("test")

	# the following call of foo will fail because it passes 
	# an int as second parameter instead of the specified str
	foo(1,2)


sphinx_typesafe/typesafe.py
###################################################################################
#
# Module containing the main typesafe decorator.
#
# Author: Klaas <khz@tzi.org> 
# Date: 03 / 2013
# Hosted on Github:  https://github.com/pythononwheels/icanhastype
#
# Maintainer: Richard Gomes<rgomes.info@gmail.com>
# See: CHANGES.rst for full list of modifications
# Hosted on Github:  https://github.com/frgomes/sphinx_typesafe
###################################################################################

from __future__ import unicode_literals
from __future__ import print_function


class typesafe(object):
    """Decorator to verify function argument types inspired and partly take from the book:
    
        Pro Python. (Expert's Voice in Open Source) from Marty Alchin

    See: http://www.amazon.com/Python-Experts-Voice-Open-Source/dp/1430227575
    """

    import re
    __types_re = re.compile(r":type[\s]+(\w+):[\s]+([\w\.]+)", re.IGNORECASE)
    __rtype_re = re.compile(r":rtype:[\s]+([\w\.]+)", re.IGNORECASE)

    __error = "Wrong type for '{}': expected: {}, actual: {}."


    def __init__(self, *args, **kwargs):
        self.noparams = len(args) == 1 and not kwargs and callable(args[0])
        if self.noparams:
            # decorator called without parameters
            self.func = args[0]
            self.types = self.inspect_function(self.func)
        else:
            # decorator called with parameters
            self.func = None
            import sys
            if sys.version_info[0] == 2:
                self.types = self.parse_params2(*args, **kwargs)
            else:
                self.types = self.parse_params3(*args, **kwargs)

    def __call__(self, *args, **kwargs):
        if self.noparams:
            # decorator called without parameters
            # --> call function stored at constructor
            self.validate_function(self.func, *args, **kwargs)
            result = self.func(*args, **kwargs)
            self.validate_result(result)
            return result
        else:
            # decorator called with parameters
            # --> return wrapper to function received in this __call__
            self.func = args[0]
            def wrapped(*a, **kw):
                self.validate_function(self.func, *a, **kw)
                result = self.func(*a, **kw)
                self.validate_result(result)
                return result
            return wrapped

    def get_class_type(self, kls):
        """
        get and return the type of a class 
        :type kls: types.StringType
        """
        import types
        #print("get_class_type for %s" % (kls), type(kls))
        if type(kls) == str:
            if kls.count('.') > 0:
                parts = kls.rpartition('.')
                import importlib
                m = importlib.import_module(parts[0])
                t = getattr(m, parts[2])
            else:
                import importlib
                m = importlib.import_module('__builtin__')
                t = getattr(m, kls)
            if t is type:
                return t
            elif isinstance(t, ( types.TypeType, 
                                 types.ClassType, 
                                 types.FunctionType )):
                return t
            else:
                return type(t)
        else:
            raise NameError("type name must be a string literal")

    def convert_entries_to_types(self, types):
        #-- print('types: ', types)
        import collections
        result = collections.OrderedDict()
        for name, atype in types:
            #print("trying to get Type %s for: %s" % (name, atype))
            obj = self.get_class_type(atype)
            #print("got: %s" % (obj))
            result[name] = obj
        #-- print('result: ', result)
        return result

    def parse_params2(self, *args, **kwargs):
        """Obtain argument types of a decorated function from parameters passed to the decorator itself.
        This behavior is only valid in Python2 paltforms.
        """
        # the parameter spec is passed as a dictionary in args[0]
        if len(args) != 1 or kwargs:
            raise AttributeError('@typesafe: illegal number of parameters')
        if isinstance(args[0], dict):
            return self.convert_entries_to_types(args[0].items())
        else:
            raise AttributeError('@typesafe: parameter must be a dictionary')

    def parse_params3(self, *args, **kwargs):
        """Obtain argument types of a decorated function by instrospecting the function itself.
        This behavior is only valid in Python3 paltforms.
        """
        raise NotImplementedError()

    def inspect_function(self, func):
        """Obtain argument types of a decorated function by instrospecting its Sphinx docstring.""" 
        import inspect
        # the parameter spec is defined as docstring.
        doc = inspect.getdoc(func)
        entries = self.__types_re.findall(doc)
        entries.append( ('return', self.__rtype_re.search(doc).group(1)) )
        return self.convert_entries_to_types(entries)

    def validate_function(self, func, *args, **kwargs):
        """Validate formal parameters before calling a decorated function."""
        import inspect
        from itertools import chain
        spec = inspect.getargspec(func)
        for name, arg in chain(zip(spec.args, args), kwargs.items()):
            # print('Check argument type ', type(arg), 'against', self.types[name])
            if name in self.types and not isinstance(arg, self.types[name]):
                raise TypeError(self.__error.format(name, self.types[name], type(arg)))

    def validate_result(self, result):
        """Validate returned value of a decorated function."""
        if 'return' in self.types:
            # print('Check returned type ', type(result), 'against', self.types['return'])
            if not isinstance(result, self.types['return']):
                raise TypeError(self.__error.format('return', self.types['return'], type(result)))
        else:
            # print('Check returned type ', type(result), 'against <NoneType>')
            if result is not None:
                raise TypeError(self.__error.format('return', 'None', type(result)))


if __name__ == "__main__":
    print("Not intended to be used on the cli ;)")
    print("Usage: see: https://github.com/frgomes/sphinx_typesafe")
